### PR TITLE
[Merged by Bors] - Implement FusedIterator for eligible Iterator types

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Most of our `Iterator` impls satisfy the requirements of `std::iter::FusedIterator`, which has internal specialization that optimizes `Interator::fuse`. The std lib iterator combinators do have a few that rely on `fuse`, so this could optimize those use cases. I don't think we're using any of them in the engine itself, but beyond a light increase in compile time, it doesn't hurt to implement the trait.

## Solution
Implement the trait for all eligible iterators in first party crates. Also add a missing `ExactSizeIterator` on an iterator that could use it.

### CODE REVIEW

File: crates/bevy_reflect/src/struct_trait.rs [hunk 1]
```diff
@@ -92,6 +92,7 @@ impl<'a> Iterator for FieldIter<'a> {
 }
 
 impl<'a> ExactSizeIterator for FieldIter<'a> {}
+impl<'a> FusedIterator for FieldIter<'a> {}
```
Comment (bjorn3) [inline crates/bevy_reflect/src/struct_trait.rs:95]: This impl may be incorrect for some `Reflect` impls.

File: crates/bevy_reflect/src/map.rs [hunk 2]
```diff
@@ -235,6 +236,7 @@ impl IntoIterator for DynamicMap {
 }
 
 impl<'a> ExactSizeIterator for MapIter<'a> {}
+impl<'a> FusedIterator for MapIter<'a> {}
```
Comment (bjorn3) [inline crates/bevy_reflect/src/map.rs:239]: Same

File: crates/bevy_reflect/src/array.rs [hunk 3]
```diff
@@ -209,6 +210,7 @@ impl<'a> Iterator for ArrayIter<'a> {
 }
 
 impl<'a> ExactSizeIterator for ArrayIter<'a> {}
+impl<'a> FusedIterator for ArrayIter<'a> {}
```
Comment (bjorn3) [inline crates/bevy_reflect/src/array.rs:213]: Same

File: crates/bevy_reflect/src/tuple.rs [hunk 4]
```diff
@@ -74,6 +75,7 @@ impl<'a> Iterator for TupleFieldIter<'a> {
 }
 
 impl<'a> ExactSizeIterator for TupleFieldIter<'a> {}
+impl<'a> FusedIterator for TupleFieldIter<'a> {}
```
Comment (bjorn3) [inline crates/bevy_reflect/src/tuple.rs:78]: Same

File: crates/bevy_reflect/src/tuple_struct.rs [hunk 5]
```diff
@@ -75,6 +76,7 @@ impl<'a> Iterator for TupleStructFieldIter<'a> {
 }
 
 impl<'a> ExactSizeIterator for TupleStructFieldIter<'a> {}
+impl<'a> FusedIterator for TupleStructFieldIter<'a> {}
```
Comment (bjorn3) [inline crates/bevy_reflect/src/tuple_struct.rs:79]: Same

**Review by bjorn3 — COMMENTED** [review]
LGTM apart from the impls in bevy_reflect.

**Review by bjorn3 — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@james7132 can you rebase?

**Comment by james7132** [comment]
@alice-i-cecile Done!

**Review by alice-i-cecile — APPROVED** [review]

**Pull request closed without merging**